File: src/leetCodeProblems/PrefixSum/BinaryTreeHasPathSum112.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/*
- Leetcode - https://leetcode.com/problems/path-sum/description/
- Time - O(n)
- Space - O(1)
*/
func hasPathSum(root *TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	targetSum -= root.Val

	if root.Left == nil && root.Right == nil {
		return targetSum == 0
	}

	if root.Left != nil && hasPathSum(root.Left, targetSum) {
		return true
	}

	if root.Right != nil && hasPathSum(root.Right, targetSum) {
		return true
	}

	return false
}

// func main() {
// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 5
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 4
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 11
// 	root.Left.Left.Left = new(TreeNode)
// 	root.Left.Left.Left.Val = 7
// 	root.Left.Left.Right = new(TreeNode)
// 	root.Left.Left.Right.Val = 2

// 	root.Right = new(TreeNode)
// 	root.Right.Val = 8
// 	root.Right.Left = new(TreeNode)
// 	root.Right.Left.Val = 13
// 	root.Right.Right = new(TreeNode)
// 	root.Right.Right.Val = 4
// 	root.Right.Right.Right = new(TreeNode)
// 	root.Right.Right.Right.Val = 1

// 	log.Println(hasPathSum(root, 26))
// }

File: src/leetCodeProblems/PrefixSum/BinaryTreeHasPathSum112.java
package leetCodeProblems.BinaryTree;

/**
 * LeetCode Problem Link - https://leetcode.com/problems/path-sum/submissions/
 *
 * InterviewBit - https://www.interviewbit.com/problems/path-sum/
 * */

public class BinaryTreeHasPathSum112 {

	public boolean hasPathSum(TreeNode root, int targetSum) {

	   if (root == null) {
           return false;
       }

       boolean ans = false;

       int subSum = targetSum - root.val;

       // This is leaf node
       if (subSum == 0 && root.left == null && root.right == null) {
           return true;
       }

       if (root.left != null) {
           ans = ans || hasPathSum(root.left, subSum);
       }

       if (root.right != null) {
           ans = ans || hasPathSum(root.right, subSum);
       }

       return ans;
   }

   public static void main(String[] args) {

    	TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        BinaryTreeHasPathSum112 ob = new BinaryTreeHasPathSum112();

        boolean output = ob.hasPathSum(root, 8);

        System.out.println(output);

   }

   static class TreeNode {

    	int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }

    }

}
